core/src/main/java/org/apache/iceberg/V4ManifestReader.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.io.CloseableGroup;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * Reader for V4 manifest files containing TrackedFile entries.
+ *
+ * <p>Supports reading both root manifests and leaf manifests. Returns TrackedFile entries which can
+ * represent data files, delete files, or manifest references. TODO: implement caching.
+ */
+class V4ManifestReader extends CloseableGroup implements CloseableIterable<TrackedFile<?>> {
+  static final ImmutableList<String> ALL_COLUMNS = ImmutableList.of("*");
+
+  private final InputFile file;
+  private final InheritableTrackedMetadata inheritableMetadata;
+  private final Long manifestFirstRowId;
+
+  private Collection<String> columns = null;
+  private TrackedFile<?> manifestDV = null;
+
+  protected V4ManifestReader(
+      InputFile file, InheritableTrackedMetadata inheritableMetadata, Long manifestFirstRowId) {
+    this.file = file;
+    this.inheritableMetadata = inheritableMetadata;
+    this.manifestFirstRowId = manifestFirstRowId;
+  }
+
+  public V4ManifestReader select(Collection<String> newColumns) {
+    this.columns = newColumns;
+    return this;
+  }
+
+  public V4ManifestReader withDeletionVector(TrackedFile<?> dv) {
+    this.manifestDV = dv;
+    return this;
+  }
+
+  public CloseableIterable<TrackedFile<?>> entries() {
```

Comment:
In v4, we are inverting the relationship between an entry and a file. Now, a tracked file contains its "entry" metadata as tracking info. The purpose of that change is to avoid needing additional methods to get the "entries" versus getting the "files".

The reader can now be simplified. It needs two read capabilities: `allFiles` and `liveFiles` (which filters out removed/deleted files). The `ClosableIterable` interface should use `liveFiles` because that's the simplest generic contract: when a manifest is read you don't return deleted files unless they are requested.